aptos-move/framework/aptos-framework/sources/nonce_validation.move

```diff
@@ -0,0 +1,236 @@
+
+    // An orderless transaction is a transaction that doesn't have a sequence number.
+    // Orderless transactions instead contain a nonce to prevent replay attacks.
+    // If the incoming transaction has the same (address, nonce) pair as a previous unexpired transaction, it is rejected.
+    // The nonce history is used to store the list of (address, nonce, txn expiration time) values of all unexpired transactions.
+    // The nonce history is used in the transaction validation process to check if the incoming transaction is valid.
+    struct NonceHistory has key {
+        // Key = sip_hash(NonceKey) % NUM_BUCKETS
+        // Value = Bucket
+        nonce_table: Table<u64, Bucket>,
+        // Used to facilitate prefill the nonce_table with empty buckets
+        // one by one using `add_nonce_bucket` method.
+        // This is the next_key to prefill with an empty bucket
+        next_key: u64,
+    }
+
+    // The bucket stores (address, nonce, txn expiration time) tuples.
+    // All the entries in the bucket contain the same hash(address, nonce) % NUM_BUCKETS.
+    // The first big ordered map in the bucket stores (expiration time, address, nonce) -> true.
+    // The second big ordered map in the bucket stores (address, nonce) -> expiration time.
+    // Both the maps store the same data, just in a different format.
+    // As the key in the first big ordered map starts with expiration time, it's easy to figure out which
+    // entries have expired at the current time. The first big ordered map helps with easy garbage collection.
+    // The second big ordered map helps with checking if the given (address, nonce) pair exists in the bucket.
+    struct Bucket has store {
+        // The first big ordered map in the bucket stores (expiration time, address, nonce) -> true.
+        nonces_ordered_by_exp_time: BigOrderedMap<NonceKeyWithExpTime, bool>,
+        // The second big ordered map in the bucket stores (address, nonce) -> expiration time.
+        nonce_to_exp_time_map: BigOrderedMap<NonceKey, u64>,
+    }
+
+    struct NonceKeyWithExpTime has copy, drop, store {
+        txn_expiration_time: u64,
+        sender_address: address,
+        nonce: u64,
+    }
+
+    struct NonceKey has copy, drop, store {
+        sender_address: address,
+        nonce: u64,
+    }
+
+    public(friend) fun initialize(aptos_framework: &signer) {
+        initialize_nonce_table(aptos_framework);
+    }
+
+    public entry fun initialize_nonce_table(aptos_framework: &signer) {
+        if (!exists<NonceHistory>(@aptos_framework)) {
+            let table = table::new();
+            let nonce_history = NonceHistory {
+                nonce_table: table,
+                next_key: 0,
+            };
+            move_to<NonceHistory>(aptos_framework, nonce_history);
+        };
+    }
+
+    fun empty_bucket(): Bucket {
+        let bucket = Bucket {
+            nonces_ordered_by_exp_time: big_ordered_map::new_with_config(0, 0, true),
```

Comment:
let's add new_with_reusable() function to big_ordered_map, same as new(), just passing true for reusable. that would make sure these are constant types ( because if they are not, it's probably better to not pass 0s here)

Reply:
Done.